README.md

```diff
@@ -1 +1,588 @@
+
+To do so, we're going to write a very naive implementation of this behavior. We're going to start with a very simple JavaScript class that efficiently calls an `update` method when our properties change. Follow along:
+
+```js
+/* We create a new class ... */
+class Batching {
+
+  /* ... that has a `requestUpdate` method */
+  async requestUpdate() {
+
+    /** 
+     * In here, we need a check to see if an update has already previously been requested.
+     * If an update is already requested, we don't want to do any unnecessary work!
+     */
+    if (!this.updateRequested) {
+
+      /**
+       * If no update was previously requested, we set this
+       * flag to `true` to avoid doing unnecessary work in
+       * case `requestUpdate` might get called another (or several) times
+       */
+      this.updateRequested = true;
+
+      /** 
+       * ... and _this_ is where the magic happens;
+       * This schedules a microtask that executes once JavaScript has
+       * finished executing, and since we guard against any other
+       * incoming calls for `requestUpdate`, we only do work once
+       */
+      this.updateRequested = await false;
+
+      /* Finally, we call our `update` method, which can then render some DOM, or do whatever */
+      this.update();
+    }
+  }
+
+  update() {
+    console.log("updating!");
+  }
+
+  /**
+   * For demonstration purposes we add some setters here,
+   * that when given a new value, will request an update
+   */
+  set a(val) {
+    this.__a = val;
+    this.requestUpdate();
+  }
+
+  set b(val) {
+    this.__b = val;
+    this.requestUpdate();
+  }
+}
+
+/* We instantiate a new instance of our class */
+const batching = new Batching();
+
+/* And we set multiple properties */
+batching.a = 1;
+batching.b = 2;
+
+/** 
+ * 🎉 The result:
+ * "updating!" is only logged to the console once because of the batching!
+ */
+```
+
+### Taking things a step further
+
+Neat! We now have a simple implementation of batching updates. We can now take things even further and provide some kind of API so users can `await` updates, like LitElement's `updateComplete` property. Take a look at the following code:
+
+```js
+class Batching {
+  constructor() {
+
+    /** 
+     * We initialize an `updateComplete` property with a new 
+     * Promise that we'll only resolve once `this.update()`
+     * has been called
+     */
+    this.updateComplete = this.__createDeferredPromise();
+  }
+
+  async requestUpdate() {
+    if (!this.updateRequested) {
+      this.updateRequested = true;
+      this.updateRequested = await false;
+      this.update();
+
+      /* When our update is, in fact, complete we resolve the Promise that was assigned to the `updateComplete` property ... */
+      this.__resolve();
+
+      /* ... And we assign a new promise to updateComplete for the next update */
+      this.updateComplete = this.__createDeferredPromise();
```

Comment:
Does lit element really pre-set an unresolved promise? I guess maybe because it `update()`s in response to `construction/attachement`? Interesting... 🤔

Reply:
worth the price of admission, right?

Reply:
yep https://github.com/Polymer/lit-element/blob/master/src/lib/updating-element.ts#L525